File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/command/execution/CommandDescriptor.java
package top.likoslupus.cellulosesz.api.command.execution;

import top.likoslupus.cellulosesz.api.command.CommandSourceKind;

import static top.likoslupus.cellulosesz.api.validation.TextChecks.requireNonBlank;

import static java.util.Objects.requireNonNull;

public record CommandDescriptor(
        String moduleId,
        String canonicalName,
        String permission,
        CommandSourceKind requiredSourceKind
) {

    public CommandDescriptor {
        moduleId = requireNonBlank(moduleId, "moduleId");
        canonicalName = requireNonBlank(canonicalName, "canonicalName");
        permission = requireNonNull(permission, "permission").trim();
        requireNonNull(requiredSourceKind, "requiredSourceKind");
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/command/execution/CommandDescriptor.kt
package top.likoslupus.cellulosesz.api.command.execution

import top.likoslupus.cellulosesz.api.command.CommandSourceKind
import top.likoslupus.cellulosesz.api.validation.requireNonBlank

@JvmRecord
public data class CommandDescriptor(
    public val moduleId: String,
    public val canonicalName: String,
    public val permission: String,
    public val requiredSourceKind: CommandSourceKind
) {

    init {
        moduleId.requireNonBlank { "moduleId" }
        canonicalName.requireNonBlank { "canonicalName" }
    }

}
